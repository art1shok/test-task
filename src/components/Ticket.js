import React from 'react';
import { RouteData } from './RouteData';

import {
  CardHead,
  CardLogo,
  Price,
  TicketWrapper
} from '../styled/Ticket.styled';

export const Ticket = ({ price, route }) => (
  <TicketWrapper>
    <CardHead>
      <Price>
        {`${price} ла`}
      </Price>
      <CardLogo src="../../../assets/images/logo.svg" alt="logo"/>
    </CardHead>
    {route && route.map((item, index) => (
        <RouteData
          way={`${item.origin} - ${item.destination}`}
          stops={item.stops}
          date={item.date}
          duration={item.duration}
          key={index}
        />
      )
    )}

  </TicketWrapper>
);

