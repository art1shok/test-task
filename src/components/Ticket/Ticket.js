import React from 'react';
import {
  CardHead,
  CardLogo,
  Price,
  TicketWrapper
} from '../../styled/Ticket.styled';
import logo from '../../assets/images/logo.svg';
import Way from '../Way/Way';

const Ticket = () => (
  <TicketWrapper>
    <CardHead>
      <Price>
        13 400 ла
      </Price>
      <CardLogo src={logo} alt="logo"/>

    </CardHead>
    <Way/>
    <Way/>

  </TicketWrapper>
);

export default Ticket;
