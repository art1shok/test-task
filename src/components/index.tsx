export { FlightTable } from './FlightTable';
export { Ticket } from './Ticket';
export { RouteData } from './RouteData';
