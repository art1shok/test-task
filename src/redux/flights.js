import { put, call, takeEvery } from 'redux-saga/effects';
import { apiService } from '../api/apiService';

const SET_ERROR = 'test-task/flights/SET_ERROR';
const GET_SEARCH_ID = 'test-task/flights/GET_SEARCH_ID';
const GET_TICKETS = 'test-task/flights/GET_TICKETS';

const initialState = {
  searchId: null,
  ticketsInfo: null,
  error: null
};

const reducer = (state = initialState, action) => {

  const { payload } = action;
  switch (action.type) {

    case GET_SEARCH_ID: {

      return {
        ...state,
        searchId: payload
      };
    }

    case GET_TICKETS: {

      return {
        ...state,
        ticketsInfo: payload
      };
    }

    case SET_ERROR: {

      return {
        ...state,
        error: payload,
      };
    }

    default:
      return state;
  }
};

export default reducer;

export const getSearchId = (data) => ({
  type: GET_SEARCH_ID,
  payload: data
});

export const getTickets = (data) => ({
  type: GET_TICKETS,
  payload: data
});

export const setError = (error) => ({
  type: SET_ERROR,
  payload: error,
});

export function * sagaWatcher () {
  yield takeEvery(GET_SEARCH_ID, sagaWorker);
}

export function * sagaWorker () {
  try {
    const result =
      yield call(apiService.getSearchId);
    const data =
      yield call(() => apiService.getTickets(result.searchId));
    yield put(getTickets(data));
  } catch (e) {
    yield put(setError(e));
  }
}
