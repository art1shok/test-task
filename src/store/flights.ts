import { put, call, takeEvery } from 'redux-saga/effects';
import { apiService } from '../api/apiService';
import {
  FlightsActionTypes,
  FlightsState,
  GetTicketsAction,
  SetErrorAction,
  TicketsInfo,
} from './flights.types';

export const SET_ERROR = 'test-task/flights/SET_ERROR';
export const GET_SEARCH_ID = 'test-task/flights/GET_SEARCH_ID';
export const GET_TICKETS = 'test-task/flights/GET_TICKETS';

const initialState: FlightsState = {
  ticketsInfo: {
    tickets: [],
    stop: true,
  },
  error: null,
};

const flightsReducer = (
  state = initialState,
  action: FlightsActionTypes,
): FlightsState => {
  switch (action.type) {
    case GET_SEARCH_ID: {
      return {
        ...state,
      };
    }

    case GET_TICKETS: {
      return {
        ...state,
        ticketsInfo: (action as GetTicketsAction).payload,
      };
    }

    case SET_ERROR: {
      return {
        ...state,
        error: (action as SetErrorAction).payload,
      };
    }

    default:
      return state;
  }
};

export default flightsReducer;

export const getSearchId = () => ({
  type: GET_SEARCH_ID,
});

export const getTickets = (data: TicketsInfo) => ({
  type: GET_TICKETS,
  payload: data,
});

export const setError = (error: any) => ({
  type: SET_ERROR,
  payload: error,
});

export function* sagaWatcher() {
  yield takeEvery(GET_SEARCH_ID, sagaWorker);
}

export function* sagaWorker() {
  try {
    const result = yield call(apiService.getSearchId);
    const data = yield call(() => apiService.getTickets(result.searchId));
    yield put(getTickets(data));
  } catch (error) {
    yield put(setError(error.message));
  }
}
