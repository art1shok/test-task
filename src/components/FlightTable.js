import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { FormGroup } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import { Ticket } from './Ticket';
// import { selectTicketsInfo } from '../redux/selectors';
import { getSearchId } from '../redux/flights';
import { ticket } from '../mocks/ticket.mock'

import {
  Background,
  ButtonContainer,
  CheckboxContainer,
  FilterContainer,
  Image,
  ImageContainer,
  StyledButton,
  StyledCheckbox,
  StyledContainer,
  StyledText,
  TicketsContainer,
} from '../styled/FlightTable.styled';

const filters = {
  none: 0,
  one: 1,
  two: 2,
  three: 3,
};

export const FlightTable = () => {
  const [sortType, setSortType] = useState('cheapest');
  const history = useHistory();
  const dispatch = useDispatch();
  // TODO: remove comments when API is working
  //const info = useSelector(selectTicketsInfo);
  const info = ticket;

  const paramsSet = useMemo(() => {
    const newParamsSet = {
      all: false,
      one: false,
      two: false,
      three: false,
      none: false,
    };

    const paramsArr = history.location.search !== ''
      ? history.location.search.substring(1)
        .split('&')
        .filter(item => item !== '')
      : [];

    paramsArr.forEach(item => {
      newParamsSet[item] = true;
    })
    return newParamsSet;

  }, [history.location.search]);

  useEffect(() => {
    dispatch(getSearchId());
  }, [dispatch]);

  const handleChange = useCallback((e) => {
    const { name } = e.target;
    let queryString = '?';

    Object.entries(paramsSet).forEach(([filter, isActive], index) => {
      if ((filter === name && !isActive) || (filter !== name && isActive)) {
        queryString += `${filter}&`;
      }
    });

    history.push({
      pathname: '/',
      search: queryString
    });
  }, [history, paramsSet]);

  const sortCheapest = (a, b) => a.price > b.price ? 1 : -1;

  const sortFastest = (a, b) => a.segments[0].duration > b.segments[0].duration ? 1 : -1;

  const filteredTickets = useMemo(() => {
    if (paramsSet['all'] || Object.values(paramsSet).every(item => !item)) {
      return info;
    }

    return info.filter((ticket) => {
        let isAppropriate = false;
        Object.entries(paramsSet).forEach(([filterName, isActive]) => {
          if (isActive && ticket.segments.every(segment => segment.stops.length === filters[filterName])) {
            isAppropriate = true;
          }
        });
        return isAppropriate;
      }
    );

  }, [info, paramsSet]);

  const sortedTickets = useMemo(() =>
      filteredTickets.sort(sortType === 'cheapest' ? sortCheapest : sortFastest),
    [filteredTickets, sortType]
  );


  const getVariant = (type) => sortType === type ? 'primary' : 'light';

  return (
    <Background>
      <StyledContainer>
        <ImageContainer>
          <Image src="../assets/images/plane.svg" alt="plane"/>
        </ImageContainer>

        <FilterContainer>
          <CheckboxContainer>
            <StyledText>
              Колличество пересадок
            </StyledText>
            <FormGroup controlId="formBasicCheckbox">
              <StyledCheckbox
                type="checkbox"
                label="Все"
                id="all"
                name="all"
                checked={paramsSet['all']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="Без пересадок"
                id="none"
                name="none"
                checked={paramsSet['none']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="1 пересадка"
                name="one"
                id="one"
                checked={paramsSet['one']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="2 пересадки"
                name="two"
                id="two"
                checked={paramsSet['two']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="3 пересадки"
                name="three"
                id="three"
                checked={paramsSet['three']}
                onChange={handleChange}
              />
            </FormGroup>
          </CheckboxContainer>
          <TicketsContainer>
            <ButtonContainer size="lg">
              <StyledButton
                variant={getVariant('cheapest')}
                onClick={() => setSortType('cheapest')}
              >
                Самый дешевый
              </StyledButton>
              <StyledButton
                variant={getVariant('fastest')}
                onClick={() => setSortType('fastest')}
              >
                Самый быстрый
              </StyledButton>
            </ButtonContainer>
            {sortedTickets && sortedTickets.slice(0, 5).map((item, index) => (
              <Ticket
                key={index}
                price={item.price}
                route={item.segments}
              />
            ))}

          </TicketsContainer>
        </FilterContainer>
      </StyledContainer>
    </Background>
  );
};
