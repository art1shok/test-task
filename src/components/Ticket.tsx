import React, {FC} from 'react';
import {RouteData} from './RouteData';

import {
  CardHead,
  CardLogo,
  Price,
  TicketWrapper
} from '../styled/Ticket.styled';
import {Segment} from "../store/flights.types";

interface Props {
  price: number;
  route: [
    Segment,
    Segment
  ]
}

export const Ticket: FC<Props> = ({price, route}) => (
  <TicketWrapper>
    <CardHead>
      <Price>
        {`${price.toLocaleString()} ла`}
      </Price>
      <CardLogo src="../../../assets/images/logo.svg" alt="logo"/>
    </CardHead>
    {route && route.map((item: any, index: number) => (
      <RouteData
        way={`${item.origin} - ${item.destination}`}
        stops={item.stops}
        date={item.date}
        duration={item.duration}
        key={index}
      />
    ))}
  </TicketWrapper>
);

