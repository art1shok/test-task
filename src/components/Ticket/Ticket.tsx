import React, { FC } from 'react';
import { RouteData } from '../RouteData/RouteData';

import {
  CardHead,
  CardLogo,
  Price,
  TicketWrapper,
} from '../../styled/Ticket.styled';
import { Segment } from '../../store/flights.types';

interface Props {
  price: number;
  route: [
    Segment,
    Segment
  ]
}

export const Ticket: FC<Props> = ({ price, route }) => (
  <TicketWrapper id="wrapper">
    <CardHead>
      <Price id="price">
        {`${price.toLocaleString()} ла` }
      </Price>
      <CardLogo
        src="../../../assets/images/logo.svg"
        alt="logo"
        id="img"
      />
    </CardHead>
    {route && route.map((item: any, index: number) => (
      <RouteData
        way={`${item.origin} - ${item.destination}`}
        stops={item.stops}
        date={item.date}
        duration={item.duration}
        key={index}
      />
    ))}
  </TicketWrapper>
);
