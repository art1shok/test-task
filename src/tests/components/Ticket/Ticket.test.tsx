import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import 'jest-enzyme';

import { Ticket } from '../../../components';
import { Segment } from '../../../store/flights.types';
import { CardLogo, Price, TicketWrapper } from '../../../styled/Ticket.styled';

configure({
  adapter: new Adapter(),
  disableLifecycleMethods: true,
});

const ticket = {
  price: 1120,
  carrier: 'AI',
  segments: [
    {
      origin: 'MAK',
      destination: 'MAV',
      date: '10/03/2017 07:29:46-0700',
      stops: [
        'JAK',
        'DAV',
      ],
      duration: 1200,
    },
    {
      origin: 'MAV',
      destination: 'MAK',
      date: '10/03/2017 07:29:46-0700',
      stops: [
        'DWV',
        'DAV',
      ],
      duration: 1890,
    },
  ],
};

describe('should render Ticket component', () => {
  let component: any = null;
  beforeEach(() => {
    component = shallow(
      <Ticket price={ticket.price} route={ticket.segments as [Segment, Segment]} />,
    );
  });
  it('Ticket component renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should contain wrapper', () => {
    const wrapper = component.find(TicketWrapper);
    expect(wrapper.length).toBe(1);
  });

  it('should contain correct format price', () => {
    const ticketPrice = '1,120 ла';
    const price = component.find(Price);

    expect(price.text()).toBe(ticketPrice.toLocaleString());
  });

  it('should contain logo', () => {
    const image = component.find(CardLogo);
    expect(image.length).toBe(1);
  });
});
