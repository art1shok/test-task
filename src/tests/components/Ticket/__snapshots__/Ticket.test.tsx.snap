// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Ticket component Ticket component renders correctly 1`] = `
<Ticketstyled__TicketWrapper>
  <Ticketstyled__CardHead>
    <Ticketstyled__Price>
      1,120 ла
    </Ticketstyled__Price>
    <Ticketstyled__CardLogo
      alt="logo"
      src="../../../assets/images/logo.svg"
    />
  </Ticketstyled__CardHead>
  <RouteData
    date="10/03/2017 07:29:46-0700"
    duration={1200}
    key="0"
    stops={
      Array [
        "JAK",
        "DAV",
      ]
    }
    way="MAK - MAV"
  />
  <RouteData
    date="10/03/2017 07:29:46-0700"
    duration={1890}
    key="1"
    stops={
      Array [
        "DWV",
        "DAV",
      ]
    }
    way="MAV - MAK"
  />
</Ticketstyled__TicketWrapper>
`;
